documentation/adr/0018-algorithm-migration-to-paged-files.md

```diff
@@ -0,0 +1,115 @@
+the use of a paged `PixelData` object.
+
+ - provides identical API for small and large datasets on small and large computers
+ - the switch between in-memory and on-disk (paged) operation is automatic as dataset grows
+ - the page size is a configuration parameter that is intended to be user configurable for the machines and datasets they’re using:
+    - a large page size will offer better performance if the whole dataset fits in a single page, 
+    - a moderate page size appears optimal for datasets too large for system RAM
+    - it is hard / impossible / undesirable to calculate optimal page size on a per-machine/per-file basis
+- implementation of a single `cut()` method (or function) that operates on an SQW object (either `this` or one passed as an argument) operating through the public API and agnostic to the size of the datasets.
+
+There are several algorithms that need to be updated to support large datasets.
+Any update must be made in such a way that algorithms that the whole of Horace continues to function as it currently does on "small" datasets
+until a point at which all algorithms are updated.
+During the migration it is also necessary that scripts using a subset of the updated algorithms
+can be created to perform end-to-end testing.
+These should not require code changes or alternate APIs.
+
+
+## Decision
+
+- A new API for SQW object will be created.
+The SQW class will to default to in memory operation (i.e. no paging)
+if a page size is not passed explicitly to the constructor
+until all code has been updated to support paged operation
+
+```matlab
+% unpaged: default page size is 0/inf
+s = sqw(filename)
+
+% paged with specified page size
+s = sqw(filename, page_size)
+```
+- a warning message will be output when operations are performed
+on dataset that require paging to enable a user to understand the performance drop
+- all algorithms will be created with a common API pattern.
+The two APIs will share a common implementation,
+the filename variant creating a (paged) SQW object
+	- `algorithm(filename, outfilename)`
+perform operation on a new SQW object that  will be explicitly created as paged
+	- `algorithm(sqw, [outfilename])`
+perform operation on existing SQW object (may be paged or unpaged)
+
+The example APIs for `cut` would be:
+```matlab
+% Pure in-memory operation (small file only)
+s = sqw(infile) 
+slice = cut(s)
+
+% Paged operation (in-memory for small file, paged for large file)
+s = sqw(infile, page_size)
+slice = cut(x) 		 % object => object
+
+% Filename based API
+slice = cut(infile)  % disk => object
+cut(infile, outfile) % disk => disk
+```
+
+
+## Consequences
+
+- The internal switch between in-memory and disk operations means
+the performance drop will happen at the point the dataset exceeds page size.
+This switch will present as a sudden, unexpected performance drop.
+Display of a console message will aid users in understanding the change.
```

Comment:
This is true, but also the fact that users will be inputting a file name rather than an SQW object, clues them in that there may be a performance drop.

Reply:
Once the whole of Horace has been updated to support large data it will be possible to call algorithms with paged `sqw` objects. The use of the filename vs object API isn't what triggers this performance drop, the existence of multiple pages of data does.

Reply:
you are forgetting the need to run algorithms in parallel and keep the same API. This means that interface should be
`result = algorithm(memory_object)`
`result = algorithm(memory_object,with_mpi)`
`result = algorithm(file_name)`
`result = algorithm(file_name,with_mpi)`
`result = algorithm(file_name,output_to_file)`
`result = algorithm(file_name,output_to_file,with_mpi)` % here result may me file-name 
file-based object in two last cases can be result but should it be....

This is also acceptable common inferface.